src/org.xtuml.bp.ui.text/src/org/xtuml/bp/ui/text/activity/OpenDeclarationAction.java

```diff
@@ -0,0 +1,267 @@
+import org.xtuml.bp.core.util.DocumentUtil;
+import org.xtuml.bp.core.util.EditorUtil;
+import org.xtuml.bp.core.util.UIUtil;
+import org.xtuml.bp.ui.text.TextPlugin;
+
+/**
+ * This action opens the declaration of model element references
+ *
+ */
+public class OpenDeclarationAction implements IEditorActionDelegate {
+	private IEditorPart targetEditor;
+
+	@Override
+	public void run(IAction arg0) {
+		ActivityEditor editor;
+		if (targetEditor instanceof ActivityEditor) {
+			editor = (ActivityEditor) targetEditor;
+		} else {
+			throw new RuntimeException("Expecting Activity editor. Found: " + targetEditor.getClass().getName());
+		}
+
+		ITextSelection selection = (ITextSelection) editor.getSelectionProvider().getSelection();
+		IDocument doc = editor.getDocumentProvider().getDocument(editor.getEditorInput());
+
+		try {
+			int startLine = selection.getStartLine() + 1;
+			int startOffset = selection.getOffset();
+			IRegion findWord = editor.findWord(doc, startOffset);
+			startOffset = findWord.getOffset();
+			int lineOffset = doc.getLineOffset(startLine - 1);
+			int offsetWithinLine = startOffset - lineOffset + 1;
+			ActivityEditorInput input = (ActivityEditorInput) editor.getEditorInput();
+			NonRootModelElement modelElement = input.getModelElement();
+			Body_c bdy = OALPersistenceUtil.getOALModelElement(modelElement);
+		    boolean parsed = false;
+		    
+			if (bdy == null) {
+				partialParse(modelElement, offsetWithinLine, doc);
+				bdy = OALPersistenceUtil.getOALModelElement(modelElement);
+				if (bdy == null) {
+					// element not found, just return
+					return;
+				}
+				parsed = true;
+			}
+			NonRootModelElement declarationElement = (NonRootModelElement) bdy.Finddeclaration(offsetWithinLine,
+					startLine);
+			if (declarationElement == null && !parsed) {
+				// If the user edited the OAL, it may need to be parsed  
+				partialParse(modelElement, offsetWithinLine, doc);
+				declarationElement = (NonRootModelElement) bdy.Finddeclaration(offsetWithinLine,
+						startLine);
+			}
+			if (declarationElement != null) {
+				if (declarationElement instanceof VariableLocation_c) {
+					VariableLocation_c location = (VariableLocation_c) declarationElement;
+				    lineOffset = doc.getLineOffset(location.getLinenumber() - 1);
+					editor.selectAndReveal(lineOffset + location.getStartposition() - 1,
+							location.getEndposition() + 1 - location.getStartposition());
+				} else {
+					boolean showInME = shouldShowInME(declarationElement);
+					if (showInME) {
+						showInModelExplorer(declarationElement);
+					} else {
+						showInModelExplorer(declarationElement);
+						Package_c pkg = declarationElement.getFirstParentPackage();
+						IEditorPart editorPart = EditorUtil.openEditorForElement(pkg);
+						editorPart.getSite().getSelectionProvider()
+								.setSelection(new StructuredSelection(declarationElement));
+						zoomToSelected(editorPart);
+					}
+				}
+			}
```

Comment:
Comment on this flow: If I understand correctly...
1. Check if there is a Body_c instance. If no, parse.
2. Search for the declaration, if found reveal it.
3. If not found, parse the body and search again, if found reveal it.

I do note that the boolean variable "parsed" prevents the body from being parsed twice. This is good. In the worst case (selecting an element that does not have a declaration e.g. a keyword, comment, symbol, etc.), the code will search and fail, parse, search and fail again.

I think this is probably acceptable since the user is accessing a feature from a CME and so instantaneous response is not as critical to UX as something like content assist. That being said, another option is to just fail after the first time and not attempting to reparse. We have established that most users always edit with "parse while editing" enabled, so we could rely on that (and obviously the case where there is no Body_c at all would still be handled). I'm not suggesting this option, just a consideration. I think the strategy we have is pretty good.